Lots of Features, Coding Standards and Fixes

mossyblog opened the pull request:
## New Entities
-  Added `Symbol.js` to Mappings Wrapper.
- Added `Fill.js` for encapsulating abstraction to `layer.fill` or `layer.fills` needs.
- Added `Border.js` for encapsulating abstraction to `layer.border` or `layer.borders` needs.
- Added `ColorHelper.js` for encapsulating RGB, HSV, HUE, SVG and HEX conversions.
- Added `Utility.js` to start the move in the direction of #6 (issuecomment-255722876)

## Feature Additions
**Selection.js**
- Added `first()` to `Selection.js` so that if you want simply the first item in the selection developers don't have to iterate through the inner workings just to resolve a API type.

**WrapperObject.js**
 - Added `wrapObject` overload ( #32 ) that points to `Application.wrapObject` that also automatically self references.
- Renamed `_object` to `sketchObject` given its usage can be used both internally and externally (plugin dev like myself need a resolver between new and old layers).

**Application.js**
-  Added `Application.reloadInspector` for situations that require the client to repaint itself. `SharedStyle` changes require this.
- Added `Application.selectedNativeDocument` to enable access to the underlying `MSDocument` DOM
- Added `Application.showDialog` to enable standardised Dialog creation.

**Document.js**
- Added  `newPage` that adds blankPages to `selectedDocument`
- Moved `wrapObject` into `WrappedObject.js` which removes a lot of the mixed singleton access points throughout the code.

**ColorHelper.js**
-  Added  `adjustBrightness` basically the utility that can add % of white/black to a HEX value (useful for designer color palette hue/sat designs).
- Added `isHex` using regex will determine if a value is a valid HEX Color code
- Added `hexToRgb` converts HEX codes to RGB struct
- Added `hexToRgbPercent` converts HEX codes to RGB Percent struct
- Added `hexToNativeColorFormat` converts hex to `MSColor`
- Added `tryGetHex` attempts to get unknown object type to `MSColor`where possible (similiar to `TryGetValue()` in other languages such as .NET)
- Added `nativeToHex` converts MSColor to HEX (ensures future ObjectiveC dependencies on `MSColor.colorWithSVGString` are in place.
- Added `hsvToRgb` converts HSV to RGB
- Added `hslToRgb` converts to HSL to RGB
- Added `rgbToHex` converts to RGB to HEX
- Added `rgbToHsv` converts to RGB to HSV
- Added `rgbToHsl` converts to RGB to HSL
- Added `NamedColor` struct which enables you to resolve "named" colors to their HEX format.

**Layer.js**
- Added `moveToPosition` to enable layers to be moved to a X/Y positions.

**Style.js**
- Added `fill` single and plural getter/setters that can take Hex Colors, `MSColor` or `MSStyleFill`
- Added `border` single and plural getter/setters that can take Hex Colors, `MSColor` or `MSStyleBorder`
-  Added `SharedStyle` Updater and Converter for turning a defined Style into a SharedStyle (need to do the removals etc next)

## Fixes / Issue Solutions

**README.md**
-  Added **Developer** section to`README.md` from Master as well as fixing the default write wasn't correct. Developers would need the .plist extension for this to work correctly. Also added `newPage` / `newArtboard` example code

**Style.js** 
- Fixed (#23) `colorFromString` was using older objectiveC methods that are no longer available to us.

**Header.js** 
- Fixed case sensitivity for `Sketch` and `Rectangle`.

**Document.js** 
- Added `wrapObject` to`layerNamed()` as if you're searching for a layer by name it didn't resolve to the API centric types which would make all the isGroup (#11) , isPage etc moot. 
-  Fixed (#22) (`shapeWithRect_` function not found)and reverted it back to the master branches approach to shape creation.
- Refactored `layerNamed` to use Predicate search to enable partial keyword search
- Added `allLayersNamed()` to enable a search on all layers with the keyword provided. Will return an array of SketchAPI objects.
- Refactored `pages` to use `wrapObject` instead of `new Page()`

**gulpfile.js** 
- Fixed `gulp` issue relating to `__globals = undefined` in _Custom Plugin_ console. `Header.js` was not being prepended to the `api.js`. Looking at the current `SketchAPI.js` it appears it's been there, so using [prependify](https://www.npmjs.com/package/prependify) I just injected it back into place before _browserify_ did its work.

## TODO

- [x]  Predicate Search for LayerNames
- [ ]  SharedStyle Conversion
- [ ]  Add/Remove Style.border and Style.fill by Fill.js/Border.js as arguments
- [ ]  Symbol Unlinking
- [ ]  Layer to Symbol conversion.
- [x]  Layers parent and reparenting.

kgn commented:
For why my opinion is worth, these updates look great!

mossyblog commented:
@kgn thanks! ... I'll be adding about 40-50 hours next week to this project in order to bring it into a position where I can write a @Unity3D export pipeline so figured I'd throw it at you folks to see if there is any value..

mossyblog commented:
- Refactored `Style.js` so that `ColorHelper.js` can take over more of the general Color duties.

- Adding both `Boder.js` and `Fill.js` to the mix. Given the intent here is to abstract SketchClient from developers my spidey senses would argue that this is necessary (having written Minecraft mods/plugins and faced the pain of dependency changes on native clients - ASM Bytecode is the true north)

- Added a pointer back to `Application.js` within `WrapperObject.js` via name of **sketchApplication**. Ideally I'd personally prefer a more IoC way of attacking this pattern, but if you want to reloadInspector or showDialog right now you'd need to know the inner workings of `NSApplication` / `NSDocumentController` which i think misses the point of this API.

mossyblog commented:
- Fixed `gulp` issue relating to `__globals = undefined` in _Custom Plugin_ console. `Header.js` was not being prepended to the `api.js`. Looking at the current `SketchAPI.js` it appears it's been there, so using `prependify` I just injected it back into place before _browserify_ did its work.

- Documented some of the new code.

kgn commented on Source/ColorHelper.js:
```diff
@@ -1,49 +1,64 @@
-  }
+    static get Black() {
+        return "#000000";
+    }
```
Why uppercase for these function names?

mossyblog commented:
Microsoft coding standards habits .. will need to go back to the old ways again heh

kgn commented:
It'd be good to have coding standards for the repo, but seems only classes are capitalized.

mossyblog commented:
Yeah I'm open to any standards imposed as whatever keeps the consistency
alive is always a positive thing.



On Sunday, 30 October 2016, David Keegan <notifications@github.com> wrote:

> It'd be good to have coding standards for the repo, but seems only classes
> are capitalized.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BohemianCoding/SketchAPI/pull/36#issuecomment-257127545>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA2qfs3Uxbad6YxHOdk2rIbAD0deUfYyks5q4_xjgaJpZM4KjXdw>
> .
>


-- 
---
Regards,
Scott Barnes
http://www.riagenic.com

mossyblog commented:
- Added more Color conversion methods
- Brought the documentation and code coverage back into a higher %  level.

kgn commented on Source/Border.js:
```diff
@@ -50,7 +50,7 @@ export class Border extends WrappedObject {
   set flatColor(value) {
     // TODO : Create internal overloader that routes color conversion based on type (hex, int etc)
-    var hexToNSColor = ColorHelper.colorFromHex(value);
+    var hexToNSColor = ColorHelper.hexToMSColor(value);
```
This just may be me, and again should be defined by coding standards. But since it's early days and we don't have this I'll insert my opinion ;) I'm not a big fan of Hungarian notation and think that the end user of the API shouldn't have to know about the underlying objects like `MSColor`, so think this would be better as simply `hexToColor`.

mossyblog replied:
Well that would be misleading as there is no "Color" type and the other
conversion methods handle the generic colour approaches. This explicitly
enables _Hex To Native_ color for when devs need an actual native
implementation. 

The framework may provide coverage for 90% of edge cases
but going in/out of the two layers should be simplified as much as possible
as well - as in where possible provide a bridge to remove deeper
dependencies.

There is NSColor and MSColor too btw

kgn commented on Source/Border.js:
```diff
@@ -112,6 +112,74 @@ export class Border extends WrappedObject {
+          "testGradientTypes" : function(tester) {
+            var border = new Border()
+
+            border.gradientType = ColorHelper.gradientTypeEnum["flat"];
```
Another opinion ;) my preference would be to remove the `Enum` from them, but to be verbose for clarity: `position` instead of `pos`. (This is my Obj-c/Swift developer mindset talking :)

```javascript
border.positions = borderPosition.outside
border.positions = borderPosition.center

ColorHelper.gradientType.flat
ColorHelper.gradientType.linear
ColorHelper.gradientType.radial
```

Also, why are the two enums used differently? It looks like `borderPosEnum` is a global object? If that's the base it probably should be capitalized: `BorderPosition`. But then why is gradient type different, being hung off of `ColorHelper`?

mossyblog replied:
I must admit I did agonise a little over whether to create a seperate class
for enums or just key/value pair dictionary style then. I choose the other
route given their footprint was small.

gradient is shared with borders and fills. I'm yet to implement the fills
and bring it all together ..

kgn commented on Source/ColorHelper.js:
```diff
@@ -9,56 +9,459 @@ import {
+    "whitesmoke": "#f5f5f5",
+    "yellow": "#ffff00",
+    "yellowgreen": "#9acd32"
+};
```
Is this useful enough to be in this API layer? This doesn't seem like something that will be used a lot, and based on this diagram if this was needed by many people it might be better in the utilities layer: https://github.com/BohemianCoding/SketchAPI/issues/6#issuecomment-255722876

mossyblog replied:
Well the helper is essentially a utility pattern. The idea here is that
I've noticed a lot of plugins export / import to CSS for variety of reasons
and the ability to convert both ways back to rgb will likely be of use to
some.

Also I'm not sure how the Utility layer can be side-loaded without having to do the same process as the API layer. In that effectively just becomes a plugin right?  As the release management would dictate you'd have to push a new API + UtilityLibrary each time for parity reasons and/or deployment reasons. Even if the API has no changes the Utility layer would still cause a iteration change.

mossyblog commented:
I must admit I did agonise a little over whether to create a seperate class
for enums or just key/value pair dictionary style then. I choose the other
route given their footprint was small.

gradient is shared with borders and fills. I'm yet to implement the fills
and bring it all together ..

On Monday, 31 October 2016, David Keegan <notifications@github.com> wrote:

> *@kgn* commented on this pull request.
> ------------------------------
>
> In Source/Border.js
> <https://github.com/BohemianCoding/SketchAPI/pull/36#pullrequestreview-6365181>
> :
>
> > +            var border = new Border()
> +
> +            border.positions = borderPosEnum["outside"];
> +            tester.assertEqual(border.positions, 2)
> +
> +            border.positions = borderPosEnum["inside"];
> +            tester.assertEqual(border.positions, 1)
> +
> +            border.positions = borderPosEnum["center"];
> +            tester.assertEqual(border.positions, 0)
> +          },
> +
> +          "testGradientTypes" : function(tester) {
> +            var border = new Border()
> +
> +            border.gradientType = ColorHelper.gradientTypeEnum["flat"];
>
> Another opinion ;) my preference would be to remove the Enum from them,
> but to be verbose for clarity: position instead of pos. (This is my
> Obj-c/Swift developer mindset talking :)
>
> border.positions = borderPosition.outsideborder.positions = borderPosition.center
> ColorHelper.gradientType.flatColorHelper.gradientType.linearColorHelper.gradientType.radial
>
> Also, why are the two enums used differently? It looks like borderPosEnum
> is a global object? If that's the base it probably should be capitalized:
> BorderPosition. But then why is gradient type different, being hung off
> of ColorHelper?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BohemianCoding/SketchAPI/pull/36#pullrequestreview-6365181>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA2qfuFDTaZadgzv2vntgfBdO9MQfEy1ks5q5LDJgaJpZM4KjXdw>
> .
>


-- 
---
Regards,
Scott Barnes
http://www.riagenic.com

mossyblog commented:
Well the helper is essentially a utility pattern. The idea here is that
I've noticed a lot of plugins export / import to CSS for variety of reasons
and the ability to convert both ways back to rgb will likely be of use to
some.

On Monday, 31 October 2016, David Keegan <notifications@github.com> wrote:

> *@kgn* commented on this pull request.
> ------------------------------
>
> In Source/ColorHelper.js
> <https://github.com/BohemianCoding/SketchAPI/pull/36#pullrequestreview-6365208>
> :
>
> > +    "slategray": "#708090",
> +    "snow": "#fffafa",
> +    "springgreen": "#00ff7f",
> +    "steelblue": "#4682b4",
> +    "tan": "#d2b48c",
> +    "teal": "#008080",
> +    "thistle": "#d8bfd8",
> +    "tomato": "#ff6347",
> +    "turquoise": "#40e0d0",
> +    "violet": "#ee82ee",
> +    "wheat": "#f5deb3",
> +    "white": "#ffffff",
> +    "whitesmoke": "#f5f5f5",
> +    "yellow": "#ffff00",
> +    "yellowgreen": "#9acd32"
> +};
>
> Is this useful enough to be in this API layer? This doesn't seem like
> something that will be used a lot, and based on this diagram if this was
> needed by many people it might be better in the utilities layer: #6
> (comment)
> <https://github.com/BohemianCoding/SketchAPI/issues/6#issuecomment-255722876>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BohemianCoding/SketchAPI/pull/36#pullrequestreview-6365208>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA2qfr59gEdPZEbKxTWAbYNDiry-FF2cks5q5LFlgaJpZM4KjXdw>
> .
>


-- 
---
Regards,
Scott Barnes
http://www.riagenic.com

mossyblog commented:
Well that would be misleading as there is no "Color" type and the other
conversion methods handle the generic colour approaches. This explicitly
enables hex to. Stove color for when devs need an actual native
implementation. The framework may provide coverage for 90% of edge cases
but going in/out of the two layers should be simplified as much as possible
as well - as in where possible provide a bridge to remove deeper
dependencies.

There is NSColor and MSColor too btw


On Monday, 31 October 2016, David Keegan <notifications@github.com> wrote:

> *@kgn* commented on this pull request.
> ------------------------------
>
> In Source/Border.js
> <https://github.com/BohemianCoding/SketchAPI/pull/36#pullrequestreview-6365170>
> :
>
> > @@ -50,7 +50,7 @@ export class Border extends WrappedObject {
>    */
>    set flatColor(value) {
>      // TODO : Create internal overloader that routes color conversion based on type (hex, int etc)
> -    var hexToNSColor = ColorHelper.colorFromHex(value);
> +    var hexToNSColor = ColorHelper.hexToMSColor(value);
>
> This just may be me, and again should be defined by coding standards. But
> since it's early days and we don't have this I'll insert my opinion ;) I'm
> not a big fan of Hungarian notation and think that the end user of the API
> shouldn't have to know about the underlying objects like MSColor, so
> think this would be better as simply hexToColor.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BohemianCoding/SketchAPI/pull/36#pullrequestreview-6365170>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA2qflkSTkfv0AqFaNP-NymrV7eLmpl0ks5q5LChgaJpZM4KjXdw>
> .
>


-- 
---
Regards,
Scott Barnes
http://www.riagenic.com

mossyblog commented:
- Incorporated ideas/changes that @kgn gave.

- Stubbed `Fill.js` to help the DOM tree retain developer affordance.

- Added `Utility.js` to start the move in the direction of #6 (issuecomment-255722876)

plrthink commented:
I'm not sure but I think this PR has so many changes that may be hard to review and test.

And may be eslint would be better?

Just my opinion.

mossyblog commented:
@plrthink Yup, and its likely to get bigger, so where is the "cut-off" point really? 

eslint which style guide - Google, Standard or Airbnb ?

Seems most of the packages for eslint in Atom aren't updated either.

plrthink commented:
@mossyblog I think we need to get some inputs from the BohemianCoding team.

mossyblog commented:
Sure, in the meantime though i'm opting for Standard can always move to a different blend later. Also note that eslint may not work out cleanly, as areas where we take dependencies on native API it will trigger warnings around them not being defined... so will need to also bake in ignores or forgiveness layers.

plrthink commented:
@mossyblog Awesome

mossyblog commented:
- Moved `wrapObject` into `WrappedObject.js` which removes a lot of the mixed singleton access points throughout the code.

- _I think  the whole factory approach to resolving API centric types here can be better handled in a more centralised approach. I had agonised a little whether to just move all that logic into WrapperObject direct but to do that each export class would have be known, which to me is a bad code smell. Instead i simply routed the calls back to Application.js (Application scope) to marshall further from there._

- Added `first()` to `Selection.js` so that if you want simply the first item in the selection developers don't have to iterate through the inner workings just to resolve a API type.

- Added `Symbol` to the mappings lookup (More to do here)

- Refactored `Application.currentDocument` to more inline with its cousin `instance.selectedDocument` only with the name `Application.selectedNativeDocument` to ensure seperate remains intact. Ideally this should go away eventually once code coverage on likely invocation points are in place (i.e. reloadInspector etc)

mossyblog commented:
Tidied up the Notes here.. as its getting to be a massive pull request in changes ...well if you factor in time...

Where is the team? :)

bomberstudios commented:
Sorry for the late reply, life's been pretty hectic at Sketch HQ lately : )

We're pretty stoked about the activity here! I'll spend some time today reviewing all the PRs, and hopefully will get this stuff merged ASAP.

Thanks for the awesome work!

mossyblog commented:
@bomberstudios no probs, I have to write a Sketch-> Unity3D plugin for work reasons and figured my efforts are better served where possible here first (assuming i didn't write crap code - which i'm not entirely ready to rule that out either hehe).

mossyblog commented:
@bomberstudios no probs, I have to write a Sketch-> Unity3D plugin for work reasons and figured my efforts are better served where possible here first (assuming i didn't write crap code - which i'm not entirely ready to rule that out either hehe).

BenAllenUK commented:
Any progress @bomberstudios and @mossyblog? Loads of good stuff in here that I know a few of us would love to use!

mossyblog commented:
I took my foot off the gas as i wasn't sure what the team were wanting to do :/ ...

fnky commented:
This is a very much needed update that adds a lot of useful features. Any notice on the progress of the PR review would be appreciated. Good work, @mossyblog — was just about to sort that out myself, but now I don't have to!

matt-curtis commented:
@bomberstudios Any updates on the team's plans? The API is pretty limited as-is.

mathieudutour commented:
I'm going to close this as the new API is completely different
